JS.md

```diff
@@ -0,0 +1,237 @@
+`Wasm.CompileError`.
+
+The asynchronous compilation is logically performed on a copy of the state of
+the given TypedArray captured during the call to `compile`; subsequent mutations
+of the TypedArray after `compile` return do not affect ongoing compilations.
+
+In the [future](FutureFeatures.md#streaming-compilation), this function can be
+extended to accept a [stream](https://streams.spec.whatwg.org), thereby enabling
+asynchronous, background, streaming compilation.
+
+## `Wasm.Module` Objects
+
+A `Wasm.Module` object represents the stateless result of compiling a
+WebAssembly binary-format module. A `Wasm.Module` logically
+contains a single [`Ast.module`](https://github.com/WebAssembly/spec/blob/master/ml-proto/spec/ast.ml#L211).
+
+### `Wasm.Module` Constructor
+
+The `Wasm.Module` constructor has the signature:
+```
+new Module(TypedArray bytes)
+```
+If the given `bytes` argument is not a
+[TypedArray object](http://tc39.github.io/ecma262/#sec-typedarray-objects),
+a `TypeError` exception is thrown.
+
+Otherwise, this function performs synchronous compilation of the TypedArray
+object:
+* The range of bytes compiled is the byte range delimited by the typed
+  array object (starting at the first element and extending `length * BYTES_PER_ELEMENT`).
+* The compiled bytes are first logically decoded into an AST according to
+[BinaryEncoding.md](BinaryEncoding.md) and then validated according to the rules
+in [spec/check.ml](https://github.com/WebAssembly/spec/blob/master/ml-proto/spec/check.ml).
+* The spec `string` values inside `Ast.module` are decoded as UTF8 as described in [Web.md](Web.md#function-names).
+* On success, a new `Wasm.Module` instance is returned that contains the
+validated [`Ast.module`](https://github.com/WebAssembly/spec/blob/master/ml-proto/spec/ast.ml#L211).
+* On failure, a new `Wasm.CompileError` is thrown.
+
+### Structured Clone of a `Wasm.Module`
+
+A `Wasm.Module` is a
+[cloneable object](https://html.spec.whatwg.org/multipage/infrastructure.html#cloneable-objects)
+which means it can be cloned between windows/workers and also
+stored/retrieved into/from an [IDBObjectStore](https://w3c.github.io/IndexedDB/#object-store).
+The semantics of a structured clone is as-if the binary source, from which the
+`Wasm.Module` was compiled, were cloned and recompiled into the target realm.
+Engines should attempt to share/reuse internal compiled code when performing
+a structured clone although, in corner cases like CPU upgrade or browser
+update, this may not be possible and full recompilation may be necessary.
+
+Given the above engine optimizations, structured cloning provides developers
+explicit control over both compiled-code caching and cross-window/worker code
+sharing.
+
+### `Wasm.Module.prototype.instantiate`
+
+The `instantiate` method has the signature:
+```
+Wasm.Instance instantiate([importObject])
+```
+If the receiver of `instantiate` is not a `Wasm.Module` instance, a `TypeError`
```

Comment:
Nit: "receiver" is not a thing. "this value" is what you're looking for, I think